### PR TITLE
Add strange pills, RandomFillSolutionComponent

### PR DESCRIPTION
**Pull request by MisterMecky**
## About the PR

Adds strange pills, which contains 20u of a randomly picked reagent. These pills will spawn randomly in maintenance closets. 

| Reagent (20u)  | Weight    | 
|--------------|-----------|
|# **Elements**|      | 
| Aluminium| 1      | 
| Carbon| 1      | 
| Chlorine | 1      | 
| Copper | 1      | 
| Fluorine | 1      | 
| Hydrogen | 1      | 
| Iodine | 1      | 
| Lithium | 1      | 
| Mercury | 1      | 
| Potassium | 1      | 
| Phosphorus | 1      | 
| Radium | 1      | 
| Silicon | 1     | 
| Sulfur | 1      | 
| Sodium | 1      | 
|# **Medicines**|      | 
| Ipecac| 3  | 
| Omnizine|2 |  
|Tricordrazine  |3 |
|# **Narcotics**|      | 
|Desoxyephedrine  |3 |
|Ephedrine  |3 |
|SpaceDrugs |5 |
|Nocturine |3 |
|MuteToxin  | 3|
|NorepinephricAcid | 3|
|# **Toxins**|      | 
|ChloralHydrate  |3 |
|Mold | 3|
|Pax  | 3|
|Toxin | 5|

Adds a new component, RandomFillSolutionComponent. Upon MapInit, this component will pick a reagent using a weighted random prototype, then add it to the entities solution.

**Media**

strangecap

https://user-images.githubusercontent.com/62370007/229344312-79d79ac4-870a-47e8-88cd-54d71c0ca49a.mp4

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl:
- add: Added strange pills, which contain a random reagent. These can be found in maintenance.

### CODE REVIEW

**Comment by vulppine** [comment]
floor pills are back baby ‼️ ‼️

**Comment by deltanedas** [comment]
youd also have to disable spiking drinks to cheese it

**Comment by vulppine** [comment]
> youd also have to disable spiking drinks to cheese it

part of the fun of floor pills is dumping everything into a cup and chugging it

**Comment by deltanedas** [comment]
yeah but then funnymen will dump them into a 1u ethanol beaker to separate out the omnizine from the nocturine

**Comment by mirrorcult** [comment]
yeah and thats funny

**Comment by deltanedas** [comment]
ok new chemistry meta that isnt cookie clicker im sold

**Comment by mirrorcult** [comment]
i dont think it'll be meta but no reason it can't be an option for aspiring maintchemists

**Comment by mirrorcult** [comment]
that said, might need to have a few more 'useless' chem strange pills

**Comment by DogZeroX** [comment]
Stuff like raw material pills

**Comment by Flareguy** [comment]
i feel like the table should just be a blacklist instead of a whitelist, part of the excitement in SS13 was being able to get literally ANYTHING from a super death chemical to a lethal dose of meth to 50u of vodka.

**Comment by MisterMecky** [comment]
Gonna open this for review, at the moment I'm not sure if having a massive list of solutions that a pill can potentially contain is a good idea? 
- Maybe I can split the list off into its own prototype and reference it. 
- Maybe I should delete the component when its done its thing?
Lemme know what I should do.

With the element pills, I was thinking of making a bottle item that contains 30u of a random element, which would spawn randomly in maint closets. If this gets merged I'll probably open a new pr for that.

**Review by 0x6273 — CHANGES_REQUESTED** [review]
Use file-scoped namespaces for new .cs files

File: Content.Server/Chemistry/Components/SolutionManager/RandomFillSolutionComponent.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    /// </summary>
+    [DataField("weight")]
+    public float Weight { get; set; }
+
+    /// <summary>
+    ///     Solution to add.
+    /// </summary>
+    [DataField("randreagents")]
+    public Solution RandReagents { get; set; }
+}
```
Comment (metalgearsloth) [inline Content.Server/Chemistry/Components/SolutionManager/RandomFillSolutionComponent.cs:38]: This should be somewhere more generic.

File: Content.Server/Chemistry/Components/SolutionManager/RandomFillSolutionComponent.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
+using Content.Shared.Chemistry.Components;
+
+namespace Content.Server.Chemistry.Components.SolutionManager;
+
+/// <summary>
+///     Fills a solution container randomly using a weighted list
+/// </summary>
+[RegisterComponent]
```
Comment (metalgearsloth) [inline Content.Server/Chemistry/Components/SolutionManager/RandomFillSolutionComponent.cs:8]: Access

File: Content.Server/Chemistry/EntitySystems/SolutionRandomFillSystem.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    public void OnRandomSolutionFillMapInit(EntityUid uid, RandomFillSolutionComponent component, MapInitEvent args)
+    {
+
+        var target = _solutionsSystem.EnsureSolution(uid, component.Solution);
+        var sumOfWeights = 0;
+
+        foreach (var picked  in component.RandomList)
+        {
+            sumOfWeights += (int) picked.Weight;
+        }
```
Comment (metalgearsloth) [inline Content.Server/Chemistry/EntitySystems/SolutionRandomFillSystem.cs:28]: This is going to have rounding issues though looks like you picked this from elsewhere.

Realistically RandomSolutionEntry should implement an interface and then a helper method for this be added to randomextensions somewhere.

File: Content.Server/Chemistry/EntitySystems/SolutionRandomFillSystem.cs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+        sumOfWeights = _random.Next(sumOfWeights);
+        Solution? randSolution = null;
+
+        foreach (var picked in component.RandomList)
+        {
+            sumOfWeights -= (int) picked.Weight;
```
Comment (metalgearsloth) [inline Content.Server/Chemistry/EntitySystems/SolutionRandomFillSystem.cs:35]: Same here on rounding when it's unneeded.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Server/Chemistry/Components/SolutionManager/RandomFillSolutionComponent.cs (hunk 2 above)
Comment (MisterMecky) [reply Content.Server/Chemistry/Components/SolutionManager/RandomFillSolutionComponent.cs]: `[RegisterComponent, Access(typeof(SolutionRandomFillSystem))]`
Like this?

Comment (metalgearsloth) [reply Content.Server/Chemistry/Components/SolutionManager/RandomFillSolutionComponent.cs]: yeah

**Comment by MisterMecky** [comment]
I've reworked this to use WeightedRandomPrototype and removed RandomSolutionEntry. May need another review?

Overall its less flexible (Cant set individual quantities, multiple reagents), but it does what it needs too. If someone wants these features they can be added later...

I've set the strange pill reagent quantity to always be 20u to reflect the recent pill changes.

File: Content.Server/Chemistry/Components/SolutionManager/RandomFillSolutionComponent.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    ///     Solution name which to add reagents to.
+    /// </summary>
+    [DataField("solution")]
+    public string Solution { get; set; } = "default";
+
+    /// <summary>
+    ///     Weighted random prototype Id. Used to pick reagent.
+    /// </summary>
+    [DataField("weightedRandomId")]
+    public string WeightedRandomId { get; set; } = "default";
```
Comment (metalgearsloth) [inline Content.Server/Chemistry/Components/SolutionManager/RandomFillSolutionComponent.cs:22]: Needs customtypeserializer and also required arg.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
Adding the solution struct was fine as long as it was in shared and reuseable but up to you how you want to do it.

**Comment by MisterMecky** [comment]
The pill sprite seems to revert back to its default sprite when out of PVS (or whatever its called), I might remove the random sprites...

**Review by metalgearsloth — APPROVED** [review]

**Pull request merged**